add new Oauth grant support (#32)
SAP/scimono: scimono-compliance-tests/Readme.md

> # SCIM 2.0 Compliance tests

Reviewer: Tests - with capital t

> A set of integration tests that are executed against external SCIM 2.0 provider implementation and verify that if it is compliant with the SCIM 2.0 standard.

Reviewer: ```suggestion
A set of integration tests that are executed against SCIM 2.0 provider implementation to verify if it is compliant with the SCIM 2.0 standard.
```

Author: done

## Prerequisite

### Required tools
- JDK Version 1.8

> - Maven - version 3.x

Reviewer: Maven is not required if we distribute binaries. I think it should be listed as an option.

### Installation
Build project **scimono-compliance-tests** with maven:

```
mvn clean install
```

## Run integration tests

Just execute the shaded executable jar you previously built located in `scimono-compliance-tests/target`: 

```

> java -jar scimono-compliance-tests-${version}-shaded.jar scim.service.url={Url of the scim endpoint of service provider}

Reviewer: List an examle here. I think that the user will immediately wonder what URL exactly they should enter. Is it /scim or /scim/ or /scim/Users?

Author: done

```

## Target System Authentication

> Some of the SCIM 2.0 Service providers require authentication. Currently SCIM compliance tests support 2 types for authentication:

Reviewer: ```suggestion
Some of the SCIM 2.0 service providers require authentication. Our SCIM compliance tests, support two types of authentication:
```

### Basic Authentication

Values for the command line arguments  `basic.auth.user` and `basic.auth.password` must be provided alongside with `auth.type=Basic` parameter. For example:
```
java -jar scimono-compliance-tests-${version}-shaded.jar scim.service.url=http://localhost:8080/scim/v2 auth.type=Basic basic.auth.user=admin basic.auth.password=password
```

### Oauth 2.0 Authentication

Currently supported oauth grant is only **Client Credentials**

> Values for the following command line arguments: `oauth.clientId`;  `oauth.secret`; `oauth.service.url` must be provided alongside with `oauth.grant=Oauth` parameter. For example:

Reviewer 2: Is is mandatory? Also, what are the possible values of "oauth.grant" parameter?

Author: Good catch! It is fixed now

```

> java -jar scimono-compliance-tests-${version}-shaded.jar scim.service.url=http://localhost:8080/scim/v2 auth.type=Oauth oauth_grant=client_credentials oauth.clientId=admin oauth.secret=secret oauth.service.url=http://localhost:8080/oauth/token

Reviewer 2: oauth.grant? (. instead of _)

Author: done

```

## Run custom tests

> Many of the SCIM 2.0 Service providers implements only some of the SCIM 2.0 features. It is possible to execute subset of all tests that are provided. You can specify which tests to be included in test execution in `.csv` file with the following format:

Reviewer 2: Can the csv file be named in any way?

Author: Yes, it is up to how it will be named. It is required only to provide a path to it in command line argument of the executable jar

> Many of the SCIM 2.0 Service providers implements only some of the SCIM 2.0 features. It is possible to execute subset of all tests that are provided. You can specify which tests to be included in test execution in `.csv` file with the following format:

Reviewer: ```suggestion
Many of the SCIM 2.0 service providers implement only some of the SCIM 2.0 features. It is possible to execute subset of all tests that are provided. You can specify which tests to be included in the test execution in `.csv` file with the following format:
```

Author: done

```
${Full class name}=${test methods separated by ','}
```

Example:
```csv
com.sap.scimono.scim.system.tests.E2EUserComplianceTest=testCreateUserRequiredAttributes, testCreateUserRequiredAttributesAndGet
com.sap.scimono.scim.system.tests.UserOperationsHttpResponseCodeTest=testGetUser200, testGetUserNonExistingId404
...
```

## Command line parameters

|Arg name| Required  | Description  |
|---|---|---|---|---|
| scim.service.url  |  Yes |  The root URL of SCIM 2.0 service provider, e.g. http://localhost:8080/scim/v2
| auth.type  |  No |  The authentication method that service provider requires. Supported values: `Basic` / `Oauth`
| oauth.grant  | No  | The grant type used used for Oauth 2.0. Supported values: `client_credentials`
| oauth.clientId  | No  | Oauth 2.0 clientId  |
| oauth.secret  | No  | Oauth 2.0 secret  |
| oauth.service.url  | No  | Oauth 2.0 service Url used to retrieve request token  |
| basic.auth.user  | No  | Username for Basic authentication  |
| basic.auth.password  | No  | Password for Basic authentication  |
| tests.file.path  | No  | Path to `.csv` file where custom subset of all tests is specified fo execution  |

## Test Results

> After tests execution finishes, log file with name `testLogs.log` is generated in the current directory from where the executable jar was started. It's content sequentially describes each action executed against the target system

Reviewer: ```suggestion
After tests execution finishes, a log file with name `testLogs.log` is generated in the current directory from where the executable jar was started. Its content sequentially describes each action executed against the target system.
```

Author: done
